### PR TITLE
Build Crackle with Crackle

### PR DESCRIPTION
**Pull request by mrm007**
This PR replaces Preconstruct with Crackle. For that, a few structural changes had to be made.

### Task orchestrator

We needed a task orchestrator, so I tested Turborepo and [Wireit]. Nx was discussed, but since we already had experience with it in Braid and other repos, it was a known quantity. Ultimately, we decided to go with Wireit because it's more lightweight and has a better API. It also has a [VS Code extension][extension] that makes the dev experience much better.

There is still a lot of untapped potential, such as [incremental builds] and [GitHub Actions caching], but I'll leave that for another PR. [Here's a comparison](https://github.com/google/wireit#related-tools) with other tools.

[Wireit]: https://github.com/google/wireit
[extension]: https://marketplace.visualstudio.com/items?itemName=google.wireit
[incremental builds]: https://github.com/google/wireit#incremental-build
[GitHub Actions caching]: https://github.com/google/wireit#github-actions-caching

### Self build

Since we don't have Preconstruct any more, we need a way to bootstrap the first build. I added a `bootstrap` package that depends on a published version of `@crackle/cli` and runs `crackle package` on all the internal packages. This is similar to how the Preconstruct monorepo works.

Now that Crackle builds internal packages, Prettier needs to ignore the built artifacts (e.g. `index.d.ts` shenanigans from [Crackle 0.21.0]). Since the Prettier proposal is still in the discussion stage (https://github.com/prettier/prettier/issues/8506 and https://github.com/prettier/prettier/issues/8048), I used `ignore-sync` which works nicely.

[Crackle 0.21.0]: https://github.com/seek-oss/crackle/releases/tag/%40crackle%2Fcore%400.21.0

#### Fallback with `unbuild`

I also configured a build script using [`unbuild`][unbuild]. It's a good fallback for when we need to use features that Crackle doesn't support yet e.g. we had a few `require` calls that don't work in ESM but unbuild shims them (all the `require` calls are gone now, by the way).

`unbuild` has minimal configuration and is driven by the `package.json` fields, while Crackle is opinionated and creates/overwrites `package.json` fields for you, there was a need to sync the two. So I added a script that creates fake entry points (dirs with `package.json`s in them) and updates the `package.json` fields that `unbuild` expects.

[unbuild]: https://github.com/unjs/unbuild

### Dev entry points (aka "[passive watcher]")

This is a powerful feature that allows you to run `crackle dev` in a package and it will create stub entries for CJS and ESM that loads code from `src` using a [TypeScript runtime]. This is great for development, but also for testing. For example, we can run `crackle dev` in `@crackle/core` and then run the tests in `@crackle/cli` and it will use the source code of `@crackle/core` instead of the compiled code.

This was a long journey in which I tried [`jiti`][jiti], [`tsx`][tsx], [`esbuild-runner`][esb] before finally settling on [`tsm`][tsm]. Some worked, some didn't, but `tsm` seems to be mature enough and has a decent [support matrix][TypeScript runtime]. [`ts-node`][ts-node] seems like a worthy alternative, but by the time I got `tsm` working, I was tired and didn't want to try another one. I might try `ts-node` in the future, but one notable difference is `ts-node` uses SWC while `tsm` uses esbuild.

[jiti]: https://github.com/unjs/jiti
[tsx]: https://github.com/esbuild-kit/tsx
[tsm]: https://github.com/lukeed/tsm
[esb]: https://github.com/folke/esbuild-runner
[ts-node]: https://github.com/TypeStrong/ts-node
[passive watcher]: https://github.com/unjs/unbuild#-passive-watcher
[TypeScript runtime]: https://github.com/privatenumber/ts-runtime-comparison

### CODE REVIEW

File: packages/core/src/plugins/rollup/dynamic-import.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import type { Plugin } from 'rollup';
+
+import type { Format } from '../../types';
+
+export const dynamicImport = (format: Format): Plugin => ({
+  name: 'crackle:resolve-dynamic-imports',
+  renderDynamicImport() {
+    if (format === 'cjs') return { left: 'import(', right: ')' };
+  },
+});
```
Comment (askoufis) [inline packages/core/src/plugins/rollup/dynamic-import.ts:10]: Is this adding support for dynamic imports to CJS files?

File: package.json [hunk 2]
```diff
@@ -3,67 +3,68 @@
...
     "@changesets/cli": "^2.25.0",
+    "@crackle-private/bootstrap": "workspace:*",
     "@crackle/cli": "workspace:*",
-    "@playwright/test": "1.22.2",
-    "@preconstruct/cli": "^2.2.2",
+    "@crackle/core": "workspace:*",
+    "@playwright/test": "^1.32.0",
     "@preconstruct/eslint-plugin-format-js-tag": "^0.2.0",
-    "@types/fs-extra": "^9.0.13",
+    "@types/fs-extra": "^11.0.1",
```
Comment (michaeltaranto) [inline package.json:46]: Should this be v10 to match the [package version]?

[package version]: https://github.com/seek-oss/crackle/pull/84/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R53

File: packages/core/build.config.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import glob from 'fast-glob';
+import { defineBuildConfig } from 'unbuild';
+
+export default defineBuildConfig({
+  entries: glob
+    .sync(['src/index.ts', 'src/entries/**/*.ts'])
+    .map((entry) => entry.replace('.ts', '')),
+  // clean: false,
+  declaration: true,
```
Comment (michaeltaranto) [inline packages/core/build.config.ts:9]:
```suggestion
  declaration: true,
```

File: packages/core/package.json [hunk 4]
```diff
@@ -99,16 +100,17 @@
...
+    "dedent": "^0.7.0",
     "ensure-gitignore": "^1.2.0",
-    "es-module-lexer": "^1.1.0",
     "esbuild": "^0.16.14",
     "eval": "^0.1.8",
     "express": "^4.18.2",
     "fast-glob": "^3.2.12",
     "fast-memoize": "^2.5.2",
-    "fs-extra": "^10.1.0",
+    "fs-extra": "^11.1.0",
```
Comment (michaeltaranto) [inline packages/core/package.json:110]: Ah it may be this package that needs the types?

File: packages/core/src/utils/dev-entry-points.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  getPackages,
+  hasDefaultExport,
+} from './entry-points';
+import { writeIfRequired } from './files';
+import { promiseMap } from './promise-map';
+
+const getHookLoader = (id: string, format: Format) => {
+  const stringify = (value: any) => JSON.stringify(value, null, 2);
+  // ! don't change this ! unbuild searches for the string and inserts shims
+  const rekwire = 'req' + 'uire';
```
Comment (askoufis) [inline packages/core/src/utils/dev-entry-points.ts:23]: wat

File: vitest.config.ts [hunk 6]
```diff
@@ -11,7 +18,11 @@ export default defineConfig({
     },
   },
   test: {
-    exclude: [...configDefaults.exclude, 'fixtures/braid-design-system/**'],
+    exclude: [
+      ...exclude,
+      // handle the circular references between configs
+      ...(integrationInclude ?? []),
+    ],
```
[the remark this answers is not in the document]
Comment (mrm007) [reply vitest.config.ts:25]: It's a circular dependency issue with Vite/`vite-node` that others have also [spent time debugging](https://discord.com/channels/885877446098964512/885880036098461716/1075988825647620096).

File: packages/core/src/plugins/rollup/dynamic-import.ts (hunk 1 above)
Comment (mrm007) [reply packages/core/src/plugins/rollup/dynamic-import.ts:10]: It's been a while (and Rollup 3 came out in the mean time), but this [might not be needed any more](https://rollupjs.org/migration/#dynamic-import-in-commonjs-output).

File: packages/core/src/utils/dev-entry-points.ts (hunk 5 above)
Comment (mrm007) [reply packages/core/src/utils/dev-entry-points.ts:23]: https://github.com/unjs/unbuild/blob/818ced7e7a33b600167e48b1d8208fe17903bb84/src/builder/plugins/cjs.ts#L17

Comment (askoufis) [reply packages/core/src/utils/dev-entry-points.ts:23]: Love a good regex

**Review by askoufis — APPROVED** [review]

**Pull request merged**